docs/kbs_protocol.md

```diff
@@ -0,0 +1,276 @@
+hash of the `tee pubkey` field needs to be included in the custom field of HW-TEE evidence and 
+signed by HW-TEE hardware. This public key is valid until the next time KBC receives this KBS's 
+attestation challenge.
+
+- `tee-evidence`
+
+The specific content of evidence related to the HW-TEE platform software and hardware in KBC's 
+environment, different `tee-evidence` formats will be defined according to the type of TEE. This 
+format is defined by the universal Attestation-Service. 
+**KBS will not analyze the content of this structure, but will directly forward it to the Attestation-Service for verification**.
+
+## Response
+
+If Attestation-Service fails to verify the evidence, the link will be disconnected by KBS.
+
+If KBS verified the token or the attestation evidence successfully, it will return a response to KBC in the following format:
+
+```json
+{
+    "response": {
+	    /* The result status of calling KBS service API, "OK" or "ERR". */
+	    "api-status": "OK",
+	    /* The output of KBS service API, needs to be encrypted by a symmetric key randomly generated by KBS. */
+	    "api-output": "",
+	    /* Symmetric key and algorithm information for encrypting `api-output`. */
+	    "crypto-annotation": {
+            "algorithm": "",
+            "symkey-length": "",
+            /* The symmetric key used to encrypt `api-output`, which is encrypted by HW-TEE's public key */
+            "enc-symkey": ""
+	    },
+	    /* The token issued by KBS to KBC. If there is a token in the request, this field will be left blank. */
+	    "token": "",
+	}
+}
+```
+
+- `api-status`
+
+The result status of calling KBS service API, `"OK"` or `"ERR"`. If it is `"ERR"`, KBS need to 
+give specific error information output in `api-output`. Please refer to [KBS service API document]() (TODO) 
+for the definition of error information of each service API.
+
+- `api-output`
+
+The output of KBS service API, it needs to be encrypted by a symmetric key randomly generated by KBS. 
+Different service APIs define different output parameters.
+
+- `crypto-annotation`
+
+The symmetric key and algorithm used to encrypt `api-output`. Because the output of KBS service 
+API may contain large data, it is inefficient to directly use HW-TEE's public key for encryption. 
+Therefore, digital envelope technology is used here, KBS generate a random symmetric key to 
+encrypt `api-output`, and then use HW-TEE's public key to encrypt the random symmetric key.
+
+- `token`
+
+If the attestation evidence sent by KBC is verified before returning the response, KBS will issue 
+a token to KBC in this field and sign it with KBS's root key. KBC can use this token to request 
+resources or services from KBS without Challenge and Attestation within the validity period of 
+the token. If there is already a token in the KBC's request, no new token will be issued and this 
+field will be left blank.
```

Comment:
When a KBC makes a request with a valid token, the KBS will need to re-evaluate the key policy based on the original result from the AS. For example, assume that the KBC requests some key from the KBS when it first connects. The KBS asks the AS to evaluate the evidence and uses the result to evaluate the key policy. The KBS gives the KBC a token for future connections. When the KBC connects a second time it provides the token. The key policy needs to be evaluated again based on the result of the original evidence. So the KBS will probably need to keep track of the original attestation result for each token. This isn't really significant for the API, but I thought it was worth mentioning.

Reply:
Yes, it is necessary to include nonce and pubkey in the attestion report. In this case, we will use `HASH(nonce + key)` as the user data of the report. If there are new important data to be included in the report in the future, it can be easily expanded.

Reply:
I think the common scenario should be (as example):
1.KBC connect to KBS with `get_key()` request for the first time, the remote attestation is launched.
2.KBS send the evidence to AS and receives the attestaion results.
3.KBS uses the attestation results to evaluate whether the sensitive Key can be deliver to the KBC.
4.If the verificaiton is passed, the KBS will send the both **KID** corresponding Key and token to KBC.
5.If the KBC connect to KBS with `get_key()` again with a **vaild token**.
6.KBS will response the sensitive Key with the corresponding **KID** directly. It needn't evaluate the attestation results again. Because the Key is binded with **KID** only.

So I think the attestation results needn't be binded with the token.

Reply:
Ok let's assume we have two keys k1 and k2 and let's assume that the AS can return three values: reject, accept1, accept2. Accept1 and accept2 require different types of evidence. For example maybe accept2 requires the guest to boot with a newer firmware version. Finally let's assume that the KBS has a key policy which says that k1 can be deployed for evidence that meets accept1 or accept2 but that key2 can only be deployed for accept2. I think this is a realistic example. Basically key2 is newer or more sensitive and the guest owner only wants to release it to VMs that have certain evidence.

Now let's assume that a VM boots with old firmware and connects to the KBS to request k1. The evidence is given to the AS, which returns accept1. The KBS sends the key and the token to the KBC. Later the KBC requests k2. Since the request has the token, the KBS doesn't have any evidence to give to the AS. The KBS should not give k2 to the KBC because the original evidence returned accept1 rather than accept2. I think the KBS needs to store the original result of the AS so that it can evaluate the key policy when it receives the second request. The presence of the token simply means that on the first request some key was able to be released but it does not guarantee the validity of future requests.

Reply:
Hi @fitzthum ,
I think this scenario should be the pre-attestation. In runtime-attestation the key is decoupled with TCB status because the key is fetched according to the KID which stored inside image layer's annotation.

**Prerequisite**: I think the key and the firmware version should be 1:1 binded. If not, it will make KBS difficult to determine which key to deliver according to the accept**X** received from AS.

> Finally let's assume that the KBS has a key policy which says that k1 can be deployed for evidence that meets accept1 or accept2 but that key2 can only be deployed for accept2. I think this is a realistic example.

About this example, I have a question and please help to check whether it's reasonable. 
The k1 can be used in both accept1(firmware version 1) and accept2(firmware version 2), and the k2 can only be used in accept2(firmware version 2). And a **VM boot image** is encrypted with **k1**.
How to launch this VM boot image inside the firmware version 2 environment? Becasue the KBS/AS will receive the evidence with firmware version 2, so it will always deliver the **k2** to KBC (because the AS will response accept2 to KBS), but the KBC needs the k1 to decrypt the boot image.

> I think the KBS needs to store the original result of the AS so that it can evaluate the key policy when it receives the second request.

I think this solution maybe can't solve the problem you mentioned, becasue the origin result of the AS is accept1. Although the KBS can evaluate the result again, it will send the k1 again.
I think a potential workable solution should be:
1.KBC send get_key() requst to KBS with a empty token. So the attestation will be launched.
2.KBC send the evidence with **older** version firmware, so AS will response accept1 to KBS, and KBS will respond key1 and a token.
3.KBC send get_key() request to KBS again, but the token will be set to empty deliberately (ignore the previous token). So a new attestation will be launched again.
4.KBC send the evidence with **newer** version firmware, so AS will response accept2 to KBS, and KBS will respond key2 and a token.

Reply:
> I think this scenario should be the pre-attestation. In runtime-attestation the key is decoupled with TCB status because the key is fetched according to the KID which stored inside image layer's annotation.

Not sure I agree with this, but let's come back to it.

To be clear I am not saying that the firmware version determines which key will be released. So if AS returns accept1 for some evidence, that doesn't guarantee that k1 will be released. The key is still determined by the kid passed in the request. Accept1 is just the response that is required from the AS to release k1. There might be several other keys that can also be released for evidence that meets accept1. For example, you might have a key for each layer of the image and they all can be released if the evidence is at some level. Similarly I might have several keys that will only be released if the AS returns accept2.

So there isn't a 1:1 binding between firmware and the key that is released. Instead there is a policy that says that certain keys can only be released when the evidence meets certain requirements. I think this is a reasonable policy for someone to have because it would allow them to have certain keys that can only be released with a newer firmware version or with a particular agent configuration (which is part of the evidence).

The problem isn't that the KBC can't get k2. It's the opposite. I think the current design would allow the KBC to request k2 (in the second request) even if it boots with old firmware. When the guest makes the first request for k1, it sends the evidence. The AS evaluates the evidence and returns accept1. Since k1 is allowed for accept1 according to the key policy, the KBS sends back k1 and the token. The KBC can then use the token to make future requests. In some cases this is fine because the future requests will have kids that have the same policy as k1. But what if either by accident or maliciously the KBC requests k2? Unless the KBS saves the result from when the KBS first evaluated the evidence, there is no way for the KBS to know that k2 should not be released.

Reply:
Yes, I think I catch your concern here and it should be reasonable. With this token mechanism, one KBC can fetch any subquently sensitive data from KBS without any futher evaluation after the orignally remote attestation is passed (get the token successfully).

But mine understanding previously about the HW-TEE environment's property are only **Trusted** or **Untrusted**. Maybe its too ideal.
- If it's **Untrusted**, the KBS will reject to send any senstitive data directly.
- If it's **Tursted**, the KBS can send any sensitive data to it. Because we think a verified Trusted HW-TEE environment won't leak any sensitive data outside it. So a KBC with legitimate token can fetch any keys (key1 and key2).

But your concern is really a good point, maybe we can have some further discussion about this in KBS's implementation proposal.